crates/flow-client/src/client.rs

```diff
@@ -46,21 +52,109 @@ impl Client {
-            pg_client,
             shard_client,
-            user_access_token,
+            user_access_token: access_token,
+            user_refresh_token: refresh_token,
         }
     }
 
+    pub async fn refresh(&mut self) -> anyhow::Result<()> {
+        // Clear expired or soon-to-expire access token
+        if let Some(_) = &self.user_access_token {
+            let claims = self.claims()?;
+
+            let now = time::OffsetDateTime::now_utc();
+            let exp = time::OffsetDateTime::from_unix_timestamp(claims.exp as i64).unwrap();
+
+            // Refresh access tokens with plenty of time to spare if we have a
+            // refresh token. If not, allow refreshing right until the token expires
+            match ((now - exp).whole_seconds(), &self.user_refresh_token) {
+                (exp_seconds, Some(_)) if exp_seconds < 60 => self.user_access_token = None,
+                (exp_seconds, None) if exp_seconds <= 0 => self.user_access_token = None,
+                _ => {}
+            }
+        }
+
+        if self.user_access_token.is_some() && self.user_refresh_token.is_some() {
+            // Authorization is current: nothing to do.
+            Ok(())
+        } else if self.user_access_token.is_some() {
+            // We have an access token but no refresh token. Create one.
+            let refresh_token = api_exec::<RefreshToken>(
+                self.rpc(
+                    "create_refresh_token",
+                    serde_json::json!({"multi_use": true, "valid_for": "90d", "detail": "Created by flowctl"})
+                        .to_string(),
+                ),
+            )
+            .await?;
+
+            self.user_refresh_token = Some(refresh_token);
+
+            tracing::info!("created new refresh token");
+            Ok(())
+        } else if let Some(RefreshToken { id, secret }) = &self.user_refresh_token {
+            // We have a refresh token but no access token. Generate one.
+
+            #[derive(serde::Deserialize)]
+            struct Response {
+                access_token: String,
+                refresh_token: Option<RefreshToken>, // Set iff the token was single-use.
+            }
+            let Response {
+                access_token,
+                refresh_token: next_refresh_token,
+            } = api_exec::<Response>(self.rpc(
+                "generate_access_token",
+                serde_json::json!({"refresh_token_id": id, "secret": secret}).to_string(),
+            ))
+            .await
+            .context("failed to obtain access token")?;
+
+            if next_refresh_token.is_some() {
+                self.user_refresh_token = next_refresh_token;
+            }
+
+            self.user_access_token = Some(access_token);
+
+            tracing::info!("generated a new access token");
+            Ok(())
+        } else {
+            anyhow::bail!("Client not authenticated");
+        }
+    }
+
+    pub fn pg_client(&self) -> postgrest::Postgrest {
```

Comment:
> There's a down-side to re-building the client from scratch every time: we will no longer re-use the underlying connection pool, and will re-dial connections every time.

Riiight, yeah. So what if `Client` kept around a single `Postgrest` client that it hands out whenever somebody calls `pg_client()` (`postgrest::Client` is a thin wrapper around an `Arc<InnerClient>` so it can be shared easily), and which only gets recreated upon refresh. In fact, we could even check if a refresh is needed inside the call to `pg_client()` so you don't even need to think about it: any time you call `.pg_client()`, you'll get a client with valid access credentials.

Ideally I'd love to wrap `Postgrest` in a wrapper that checks and auto-refreshes its access token every time you make an API call. That way it could modify its own header map and not need to get re-created at all. But I think my proposed solution above is a decent compromise

Reply:
I think it's far too much coupling between `Client`, which we should seek to keep very un-opinionated, with a highly opinionated refresh flow.

For example, refresh() will create a refresh token if it doesn't already have one, even if there's a perfectly valid access token available. Is that appropriate for flowctl? I think "yes", though even that is possibly debatable. Is it _always_ appropriate? Definitely not.

Or what if we introduce a new mechanism for refresh tokens? We easily could -- we're currently ripping out gateway_auth_token after all.

I don't think Client should know anything about the manner in which tokens are obtained, and that that business logic belongs in free function(s) that operate _on_ a Client.

Reply:
Is [this](https://github.com/estuary/flow/pull/1665/commits/a69e6593469b8c04189fe3d67080c52a15f2831b) what you meant? I think I got a bit lost between `flow_client::Client` and `postgrest::Postgrest` when reading your comments, but this should:

* Move opinionated refresh logic to bare function to avoid coupling `flow_client::Client` to specific token refresh logic
* Create a single `postgrest::Postgrest` at `flow_client::Client` creation time and hand out clones of it with auth headers set to the current access token upon request. This should allow all users of `pg_client()` to share a connection pool, while still allowing `flow_client::Client` to hand out `postgrest::Postgrest` instances to all of the places that need them

Reply:
Back up a step, and look at how flowctl on `master` updates it's configuration and then writes it out on exit. This change, as is, breaks that workflow because token refresh is treated as in interior detail of `flow_client::Client`, and I'm saying: it shouldn't be.

`flow_client::Client` should be immutable _exactly because_ of the kind of bug which currently exists in `flowctl` within this branch.

This is how I think it should be structured:
* `flow_client::Client`, once created, is immutable. We don't do any interior mutation, ever.
* Have a `flow_client::refresh_authorizations(client: &Client, access_token: Option<String>, refresh_token: Option<String>) -> anyhow::Result<(String, String)>` routine which takes an optional access and/or refresh token, and returns a guaranteed access and refresh token (or an error).
  * This routine should be the _only_ place we parse the access token and do an expiration check -- currently this logic is also still duplicated in flowctl/lib.rs.
  * Feel free to implement a further subroutine for getting an access token from a refresh token, if that's useful for dekaf.
* _Use_ this method in `flowctl` to update the configured access and refresh token, and to then re-build the client, as it had been doing prior to this PR.

IIRC there isn't a reason for a `flow_client::Client::claims()` routine aside from checking expiration, but I might be missing something. In any case, IMO this too should be a free function if it's still needed and Client should not have an opinion on the shape of an access token.

Reply:
I see the problem, and why the raw access/refresh tokens need to be exposed -- in order for them to get set on `flowctl::Config` + written back to disk. I believe [this](https://github.com/estuary/flow/pull/1665/commits/f9fcd6cfe50e2004b398e4eb2b84092944742a3c) fixes `flowctl` writing updated creds back to disk. I tested it locally by removing `user_access_token` from my `local.json` profile, ran `cargo run -p flowctl auth roles list` and it wrote back a new access token

Reply:
Yep, that works :+1: